### PR TITLE
Modernise application.spec.ts

### PR DESCRIPTION
**Pull request by samwho**
## Description

I starting working on a bug to do with custom roles and user groups, which led me to writing some initial tests, which led me to `application.spec.ts` (because the problem is in the `fetch` for applications, I think).

I noticed these tests were written using the old helpers, and the code they were testing was missing types in a lot of places. This PR is an effort to improve the situation.

No business logic changes are in this PR, it's entirely cleanup and typing.

### CODE REVIEW

File: packages/server/scripts/test.sh [hunk 1]
```diff
@@ -3,12 +3,12 @@ set -e
 
 if [[ -n $CI ]]
 then
-  export NODE_OPTIONS="--max-old-space-size=4096 --no-node-snapshot"
+  export NODE_OPTIONS="--max-old-space-size=4096 --no-node-snapshot $NODE_OPTIONS"
```
Comment (samwho) [inline packages/server/scripts/test.sh:6]: This allows VSCode's "auto attach" feature to work. It passes values into `NODE_OPTIONS` that allows it to attach a debugger to any node process started from a VSCode terminal.

File: packages/server/src/api/controllers/application.ts [hunk 2]
```diff
@@ -196,17 +201,19 @@ export async function fetchAppDefinition(ctx: UserCtx) {
...
-  let application = await db.get<any>(DocumentType.APP_METADATA)
+  let application = await db.get<App>(DocumentType.APP_METADATA)
   const layouts = await getLayouts()
   let screens = await getScreens()
   const license = await licensing.cache.getCachedLicense()
 
   // Enrich plugin URLs
   application.usedPlugins = objectStore.enrichPluginURLs(
-    application.usedPlugins
+    application.usedPlugins || []
```
Comment (samwho) [inline packages/server/src/api/controllers/application.ts:216]: `usedPlugins` was not a member of the `App` type, so you'll notice I've added it. Slightly uncertain about that change, but without it this code doesn't type-check correclty.

File: packages/server/src/api/routes/application.ts [hunk 3]
```diff
@@ -4,7 +4,6 @@ import * as deploymentController from "../controllers/deploy"
 import authorized from "../../middleware/authorized"
 import { permissions } from "@budibase/backend-core"
 import { applicationValidator } from "./utils/validators"
-import { importToApp } from "../controllers/application"
```
Comment (samwho) [inline packages/server/src/api/routes/application.ts:7]: Unused.

File: packages/worker/scripts/test.sh [hunk 4]
```diff
@@ -4,10 +4,10 @@ set -e
 if [[ -n $CI ]]
 then
   # Running in ci, where resources are limited
-  echo "jest --coverage --maxWorkers=2  --forceExit --bail"
-  jest --coverage --maxWorkers=2  --forceExit --bail
+  echo "jest --coverage --maxWorkers=2  --forceExit --bail $@"
```
Comment (samwho) [inline packages/worker/scripts/test.sh:7]: Allows you to run `yarn test <filename>`

File: qa-core/src/internal-api/tests/tables/tables.spec.ts [hunk 5]
```diff
@@ -13,17 +13,6 @@ describe("Internal API - Table Operations", () => {
...
-  async function createAppFromTemplate() {
-    return config.api.apps.create({
-      name: generator.word(),
-      url: `/${generator.word()}`,
-      useTemplate: "true",
-      templateName: "Near Miss Register",
-      templateKey: "app/near-miss-register",
-      templateFile: undefined,
-    })
-  }
```
Comment (samwho) [inline qa-core/src/internal-api/tests/tables/tables.spec.ts:25]: Unused.

File: packages/server/src/api/controllers/application.ts [hunk 6]
```diff
@@ -174,14 +177,16 @@ export const addSampleData = async (ctx: UserCtx) => {
   ctx.status = 200
 }
 
-export async function fetch(ctx: UserCtx) {
+export async function fetch(ctx: UserCtx<null, App[]>) {
```
[the remark this answers is not in the document]
Comment (mike12345567) [reply packages/server/src/api/controllers/application.ts:180]: `void` is generally preferred if its the lack there-of we expect - its more specific in terms of meaning absolutely nothing/not there - very similar to `undefined` but stricter.

File: packages/server/src/api/controllers/application.ts [hunk 7]
```diff
@@ -47,6 +47,9 @@ import {
   PlanType,
   Screen,
   UserCtx,
+  CreateAppRequest,
+  FetchAppDefinitionResponse,
+  type FetchAppPackageResponse,
```
Comment (mike12345567) [inline packages/server/src/api/controllers/application.ts:52]: What is the need for the `type` keyword here?

File: packages/server/src/tests/utilities/api/application.ts [hunk 8]
```diff
@@ -1,17 +1,96 @@
...
+    const result = await this.request
+      .post(`/api/applications/${appId}/sync`)
+      .set(this.config.defaultHeaders())
+      .expect("Content-Type", /json/)
+      .expect(statusCode)
+    return result.body
+  }
+
   getRaw = async (appId: string): Promise<Response> => {
+    // While the appPackage endpoint does take an :appId parameter, it doesn't
```
Comment (mike12345567) [inline packages/server/src/tests/utilities/api/application.ts:85]: We probably should update this (not in this PR though!) to use the `appId` in the URL if no header is provided.

File: packages/types/src/documents/app/app.ts [hunk 9]
```diff
@@ -73,3 +75,30 @@ export interface AppFeatures {
 export interface AutomationSettings {
   chainAutomations?: boolean
 }
+
+export interface CreateAppRequest {
```
Comment (mike12345567) [inline packages/types/src/documents/app/app.ts:79]: Generally we separate the `document` definitions and the API request definitions, these should probably be moved to the `types/src/api/web` layer.

**Review by mike12345567 — APPROVED** [review]
LGTM - would update the `UserCtx<void` instances for consistency, and move the API typing into the API layer, but massive improvement to this test, much more readable, and nice to have this stuff better typed!

File: packages/server/src/api/controllers/application.ts (hunk 7 above)
Comment (samwho) [reply packages/server/src/api/controllers/application.ts]: This is something my editor seems to be when it auto-imports a type. Will look into configuring it not to do it.

File: packages/backend-core/src/objectStore/buckets/plugins.ts [hunk 10]
```diff
@@ -6,7 +6,7 @@ import { Plugin } from "@budibase/types"
 
 // URLS
 
-export function enrichPluginURLs(plugins: Plugin[]) {
+export function enrichPluginURLs(plugins: Plugin[] | undefined): Plugin[] {
```
[the remark this answers is not in the document]
Comment (samwho) [reply packages/backend-core/src/objectStore/buckets/plugins.ts:9]: 🙈

File: packages/server/src/api/controllers/application.ts [hunk 11]
```diff
@@ -58,23 +61,23 @@ import * as appMigrations from "../../appMigrations"
 async function getLayouts() {
   const db = context.getAppDB()
   return (
-    await db.allDocs(
+    await db.allDocs<Layout>(
```
Comment (adrinr) [inline packages/server/src/api/controllers/application.ts:64]: 😍

**Review by adrinr — APPROVED** [review]
Thx for the types!

**Pull request merged**